python/mxnet/test_utils.py

```diff
@@ -552,7 +552,10 @@ def random_projection(shape):
         assert isinstance(grad_stype_dict, dict), "grad_stype_dict must be a dict"
         for k, v in grad_stype_dict.items():
             if k in args_grad and v in _STORAGE_TYPE_STR_TO_ID and v != 'default':
-                args_grad[k] = mx.nd.cast_storage(args_grad[k], stype=v)
+                # create an uninitialized sparse ndarray for executor
+                # if the symbolic grad is expected to be zero, it should not be initialized at all
+                args_grad[k] = mx.nd.zeros(args_grad[k].shape, args_grad[k].context,
+                                           args_grad[k].dtype, v)
 
```

Comment:
Yes, it is the correct way to do it. The previous version is not correct. The args_grad are outputs in backward FComputeEx functions. They must be CheckAndAlloc inside the FComputeEx functions. The cast_storage would actually create an initialized sparse tensor for the backward which is not right. One corner case illustrating this is that, if the input is uninitialized, the square_sum backward function returns immediately, so the output should be uninitialized as well.